pkg/kubecost/allocation.go

```diff
@@ -1194,140 +1177,120 @@ func (a *Allocation) generateKey(properties Properties) string {
+		case agg == string(AllocationControllerKindProp):
+			controllerKind := a.Properties.ControllerKind
+			if controllerKind == "" {
+				// Indicate that allocation has no controller
+				controllerKind = UnallocatedSuffix
 			}
-		}
-
-		controller, err := a.Properties.GetController()
-		if err != nil {
-			// Indicate that allocation has no controller
-			controller = UnallocatedSuffix
-		}
-
-		names = append(names, controller)
-	}
-
-	if properties.HasPod() {
-		pod, _ := a.Properties.GetPod()
-		names = append(names, pod)
-	}
-
-	if properties.HasContainer() {
-		container, _ := a.Properties.GetContainer()
-		names = append(names, container)
-	}
-
-	if properties.HasService() {
-		services, err := a.Properties.GetServices()
-		if err != nil {
-			// Indicate that allocation has no services
-			names = append(names, UnallocatedSuffix)
-		} else {
-			if len(services) > 0 {
+			if aggControllerKind != "" && aggControllerKind != controllerKind {
+				// The allocation does not have the specified controller kind
+				controllerKind = UnallocatedSuffix
+			}
+			names = append(names, controllerKind)
+		case agg == string(AllocationControllerProp):
+			if indexOf(string(AllocationControllerKindProp), aggregateBy) == -1 &&
+				a.Properties.ControllerKind != "" {
+				names = append(names, a.Properties.ControllerKind)
+			}
+			controller := a.Properties.Controller
+			if controller == "" {
+				// Indicate that allocation has no controller
+				controller = UnallocatedSuffix
+			}
+			names = append(names, controller)
+		case agg == string(AllocationPodProp):
+			names = append(names, a.Properties.Pod)
+		case agg == string(AllocationContainerProp):
+			names = append(names, a.Properties.Container)
+		case agg == string(AllocationServiceProp):
+			services := a.Properties.Services
+			if services == nil || len(services) == 0  {
+				// Indicate that allocation has no services
+				names = append(names, UnallocatedSuffix)
+			} else {
+				// This just uses the first service
 				for _, service := range services {
 					names = append(names, service)
 					break
 				}
-			} else {
-				// Indicate that allocation has no services
-				names = append(names, UnallocatedSuffix)
 			}
-		}
-	}
-
-	if properties.HasAnnotations() {
-		annotations, err := a.Properties.GetAnnotations()
-		if err != nil {
-			// Indicate that allocation has no annotations
-			names = append(names, UnallocatedSuffix)
-		} else {
-			annotationNames := []string{}
-
-			aggAnnotations, _ := properties.GetAnnotations()
-			for annotationName := range aggAnnotations {
-				if val, ok := annotations[annotationName]; ok {
-					annotationNames = append(annotationNames, fmt.Sprintf("%s=%s", annotationName, val))
-				} else if indexOf(UnallocatedSuffix, annotationNames) == -1 { // if UnallocatedSuffix not already in names
-					annotationNames = append(annotationNames, UnallocatedSuffix)
+		case strings.HasPrefix(agg, "controllerKind:"):
+			continue
```

Comment:
This is a little confusing to read because it relies on a separate block above, but it's not exactly obvious how/why unless you understand the design of that property.